### ISSUE TITLE
Feature Request: "no" versions of boolean options

### ISSUE THREAD

**Issue by adamsmd** [remkop/picocli#694]
Some Unix commands support "--option" to turn a boolean option on and "--no-option" to turn it off.  I would like a convenient way to specify these types of options.

One possible syntax for this would work like the following:

```
@Option(names={"--foo", "-f"}, inverses={"--no-foo", "+f"})
boolean foo = readDefaultFromEnvironment("FOO");
```

**Comment by remkop** [comment]
Interesting idea.  I have seen this option style before, so this is not uncommon. 

Note that this can already be accomplished with the current version of picocli: have two `@Option`-[annotated methods](https://picocli.info/#option-parameters-methods), one that sets the field to `true`, and the other one (with the inverse option names) that sets the field to `false`.

An alternative is to allow users to specify a `true|false` value on the command line by defining the option with `arity = "1"` (users must specify a "true" or "false" option parameter), or `arity = "0..1"` (the option optionally takes a parameter).

But I'm open to considering library support for this, so let's explore:
* What would the synopsis of a command with such options look like? Does it show `-f` only, or both `-f` and `+f`, or is the inverse omitted from the synopsis? If there are multiple options like this, are they lumped together in the synopsis? E.g. short options are usually lumped together in the synopsis like `<cmd> -abc`. Would this make sense: `<cmd> -abc +abc`?
* How would the usage help message look like for such options? Do the inverse options need a separate entry with their own description?

**Comment by remkop** [comment]
Doing some research on the web, I found that Ruby's OptionParser has this functionality built in.
It looks like this only works for long options though, and is [based on the option name](https://ruby-doc.org/stdlib-2.6.1/libdoc/optparse/rdoc/OptionParser.html#method-i-def_option) (see `make_switch` source code):

* if the option name matches regex `^--no-([^\[\]=\s]*)(.+)?`, the parser assumes this is a boolean option, and assigns `false` when it is matched on the command line without option parameter.
* if the option name matches regex `^--\[no-\]([^\[\]=\s]*)(.+)?`, the parser assigns `false` if the `--no-xxx` option is matched, and `true` if the `--xxx` option is matched without option parameter on the command line.

So, adding `[no-]` to the start of the long option name makes the option "negatable".

**Comment by remkop** [comment]
Finally, in the Java world, there are the `-XX` JVM options:

> Boolean options are turned on with `-XX:+<option>` and turned off with `-XX:-<option>`.

These options are not shown on `--help`, so there is no example of what the synopsis or options list should look like for such options. Html documentation shows the default value, something like this: 

```
-XX:+AggressiveOpts
-XX:+PrintGCDetails
-XX:+UseG1GC
```

Graal's `native-image` tool has a similar style, but uses the `±` character (\u00b1) in the options list:

```
./native-image --expert-options
  -H:±AllowIncompleteClasspath                 Allow image building with an incomplete class path: report type resolution errors at run time when they are accessed the
                                               first time, instead of during image building. Default: - (disabled).
  -H:±AllowVMInspection                        Enables features that allow the VM to be inspected during runtime. Default: - (disabled).
  -H:±RuntimeAssertions                        Enable or disable Java assert statements at run time. Default: - (disabled).
  -H:±JNIVerboseLookupErrors                   Report information about known JNI elements when lookup fails. Default: - (disabled).
```

**Comment by remkop** [comment]
# Conclusion
Overall, I think this will be a nice feature to support in the library.

# Thinking out loud
Remaining questions:
* what API do we offer? 
* what does the synopsis look like?
* what does the option list look like?
* what does the reported default value in the usage help look like?

The API needs to be flexible enough to support Java-style `-XX: ±<option>` negatable options as well as unix-style `--[no-]option` for long options and `±x` for short options.

One idea is your suggestion to have two sets of names:

```java
@Option(names = {"-x", "--long-option", "-H:-java-style"}, 
      negated = {"+x", "--no-long-option", "-H:+java-style"})
boolean x;
```

This gives the parser enough information to match command line options and assign the correct boolean value. The synopsis and option list in the usage help message is a bit trickier, because we want to collapse the option names and their negated names into a single representation where possible:

```
Usage: <cmd> [-yz] [±x]
  ±x, --[no-]long-option, -H:±java-style   Enables some mode. Default: true.
```
(TBD: Should negated short options be shown in the synopsis? If there are multiple, should they be clustered? (E.g. `±abcx`))

The problem with this solution is what do we do if there is no similarity between the option names and the negated names. For example:

```java
@Option(names = {"-x", "--long-option", "-H:-java-style"}, 
      negated = {"+y", "--disallow-long-or-any-option", "-H:+java-ish-style"})
boolean x;
```

One idea is to show such negated options on a separate line under the normal options, perhaps with some auto-generated description:
```
Usage: <cmd> [-abc] [-x|+y]
  -x, --long-option, -H:+java-style   Enables some mode. Default: true.
  +y, --disallow-long-or-any-option, -H:-java-ish-style
                                      (Negated form of -x, --long-option, -H:-java-style)
```

This needs a bit more thought.

**Comment by remkop** [comment]
An alternative is to have a simple annotation API, like this:

```java
@Option(names = {"-x", "--long-option"}, negatable = true)
boolean x;
```

Then allow applications to customize the option name transformation:

```java
class RegexTransformer implements INegativeTransformer {
    Map<Pattern, String> replacements = new LinkedHashMap<Pattern, String>();
    Map<Pattern, String> synopsis = new LinkedHashMap<Pattern, String>();

    public RegexTransformer addPattern(String regex, String negativeReplacement, String synopsisReplacement) {
        Pattern pattern = Pattern.compile(regex);
        replacements.put(pattern, negativeReplacement);
        synopsis.put(pattern, synopsisReplacement);
        return this;
    }

    @Override
    public String makeNegative(String optionName) {
        for (Map.Entry<Pattern, String> entry : replacements.entrySet()) {
            Matcher matcher = entry.getKey().matcher(optionName);
            if (matcher.find()) { return matcher.replaceAll(entry.getValue()); }
        }
        return optionName;
    }

    @Override
    public String makeSynopsis(String optionName) {
        for (Map.Entry<Pattern, String> entry : synopsis.entrySet()) {
            Matcher matcher = entry.getKey().matcher(optionName);
            if (matcher.find()) { return matcher.replaceAll(entry.getValue()); }
        }
        return optionName;
    }
}

@Test
public void testRegex() {
    RegexTransformer transformer = new RegexTransformer()
        .addPattern("^-(\\w)$", "+$1", "±$1")
        .addPattern("^--(\\w(-|\\w)*)$", "--no-$1", "--[no-]$1")
        .addPattern("^(-|--)(\\w*:)\\+(\\w(-|\\w)*)$", "$1$2-$3", "$1$2±$3")
        .addPattern("^(-|--)(\\w*:)\\-(\\w(-|\\w)*)$", "$1$2+$3", "$1$2±$3")
        ;

    assertEquals("-X:-option", transformer.makeNegative("-X:+option"));
    assertEquals("-X:±option", transformer.makeSynopsis("-X:+option"));
    assertEquals("-X:+option", transformer.makeNegative("-X:-option"));
    assertEquals("-X:±option", transformer.makeSynopsis("-X:-option"));

    assertEquals("+x", transformer.makeNegative("-x"));
    assertEquals("±x", transformer.makeSynopsis("-x"));

    assertEquals("--no-verbose", transformer.makeNegative("--verbose"));
    assertEquals("--[no-]verbose", transformer.makeSynopsis("--verbose"));
}
```

**Comment by remkop** [comment]
Related: https://github.com/cbeust/jcommander/issues/412

**Comment by remkop** [comment]
@adamsmd This feature has been included in the 4.0.0-beta-1b release that just came out.
Thanks for raising this!
Please give it a try and let me know what you think.

**Comment by adamsmd** [comment]
Awesome! Looks good!
